File: Day12/getStrongest.cpp
vector<int> getStrongest(vector<int>& arr, int k) {
    sort(begin(arr),end(arr));
    int m = arr[(arr.size()-1)/2];
    unordered_map<int,vector<int>> ratingNums;
    unordered_set<int> r;
    for (auto& a : arr){
        int rate = abs(a - m);
        ratingNums[rate].push_back(a);
        r.insert(rate);
    }
    vector<int> rs(r.begin(), r.end());
    sort(begin(rs),end(rs),greater<int>());
    vector<int> ans;
    for (int i = 0; i < rs.size(); i++){
        int rate = rs[i];
        sort(ratingNums[rate].begin(), ratingNums[rate].end(), greater<int>());
        for (int j = 0; j < ratingNums[rate].size() && ans.size() < k; j++){
            ans.push_back(ratingNums[rate][j]);
        }
    }
    return ans;
}

File: Day12/getStrongest.java
public int[] getStrongest(int[] arr, int k) {
    int n = arr.length;
    Arrays.sort(arr);
    int med = arr[((n - 1) / 2)];
    TreeMap<Integer, List<Integer>> map = new TreeMap<>();
    for(int i=0;i<arr.length;i++) {
        int key = Math.abs(arr[i]-med);
        List<Integer> lst = map.getOrDefault(key, new ArrayList<>());
        lst.add(arr[i]);
        map.put(key, lst);
    }
    int[] ret = new int[k];
    int a = 0;
    while(a<k && !map.isEmpty()) {
        List<Integer> lst = map.get(map.lastKey());
        Collections.sort(lst);
        Collections.reverse(lst);
        int t = a;
        for(int i=0;i<Math.min(lst.size(),k-t);i++) {
            ret[a] = lst.get(i);
            a++;
        }
        map.remove(map.lastKey());
    }
    return ret;
}